core/src/main/scala/pureconfig/ConfigSource.scala

```diff
@@ -0,0 +1,320 @@
+ * to `ConfigSource`.
+ *
+ * Other PureConfig modules may provide other ways or building config sources (e.g. for different
+ * config formats or data sources).
+ */
+object ConfigSource {
+
+  /**
+   * A config source for the default loading process in Typesafe Config. Typesafe Config stacks
+   * `reference.conf` resources provided by libraries, application configs (by default
+   * `application.conf` in resources) and system property overrides, resolves them and merges them
+   * into a single config. This source is equivalent to
+   * `defaultOverrides.withFallback(defaultApplication).withDefault(defaultReference)`.
+   */
+  val default = ConfigObjectSource(ConfigFactoryWrapper.load())
+
+  /**
+   * A config source for the default loading process in Typesafe Config with a custom application
+   * config source. Typesafe Config stacks `reference.conf` resources provided by libraries, the
+   * given file and system property overrides, resolves them and merges them into a single config.
+   *
+   * This method is provided here to support use cases that previously depended on
+   * `ConfigFactory.load(config)`. Creating a custom source by merging the layers manually is
+   * usually recommended as it makes the config priorities more transparent.
+   *
+   * @param appSource the source providing the application config
+   * @return a `ConfigObjectSource` for the default loading process in Typesafe Config with a
+   *         custom application config source.
+   */
+  def default(appSource: ConfigObjectSource): ConfigObjectSource =
+    ConfigObjectSource(appSource.value().right.flatMap { cv => ConfigFactoryWrapper.load(cv.toConfig) })
+
+  /**
+   * A config source that always provides empty configs.
+   */
+  val empty = ConfigObjectSource(Right(ConfigFactory.empty))
+
+  /**
+   * A config source for the default reference config in Typesafe Config (`reference.conf`
+   * resources provided by libraries). Like Typesafe Config, it provides an empty object if
+   * `reference.conf` files are not found.
+   *
+   * As required by
+   * [[https://github.com/lightbend/config/blob/master/HOCON.md#conventional-configuration-files-for-jvm-apps the HOCON spec]],
+   * the default reference files are pre-emptively resolved - substitutions in the reference config
+   * aren't affected by application configs.
+   */
+  val defaultReference = ConfigObjectSource(ConfigFactoryWrapper.defaultReference())
+
+  /**
+   * A config source for the default reference config in Typesafe Config (`reference.conf`
+   * resources provided by libraries) before being resolved. This can be used as an alternative
+   * to `defaultReference` for use cases that require `reference.conf` to depend on
+   * `application.conf`. Like Typesafe Config, it provides an empty object if `reference.conf`
+   * files are not found.
+   */
+  val defaultReferenceUnresolved = resources("reference.conf").optional
```

Comment:
Good catch 👍 I'll add test cases for `defaultReferenceUnresolved` and fix the issue.

We can add a method to explicitly resolve configs, sounds like it would be useful in some cases. However, I'm still fond of the approach "keep all configs unresolved by default, force resolve them when you call and load* method". Most users aren't even aware that such a complex system of variable resolution exists, and people just calling `ConfigSource.file(myFile).load[Conf]` should have this simply working.

I agree we should move the auto-resolution part to `load` and `loadOrThrow`, to make it happen on the outer layer of the API. However, this still means that `ConfigSource.defaultReferenceUnresolved.load[A]` *will* resolve the config before it tries to apply the `ConfigReader`, which seems reasonable - at this point we are creating the config tree for the whole purpose of reading it, so having undefined references at this point doesn't seem desirable in any case.

Reply:
Fixed the existing test case in 966053f (I had the two config lines swapped) and fixed it in 74d4ef2, but not in the way we discussed. After delving a bit more into Typesafe Config's API, I noticed that:
- Only `Config` has a `resolve()` method, not `ConfigValue`;
- The [Javadoc](https://static.javadoc.io/com.typesafe/config/1.3.4/com/typesafe/config/Config.html#resolve--) of `resolve()` says that "a given Config must be resolved before using it to retrieve config values". Technically this is not entirely true as we can pass non-default `ConfigResolveOptions` that allow unresolved configs, but it implies that `ConfigValue`s are not supposed to be unresolved at any point;
- People can easily convert between `Config` and `ConfigObject`, but forcing resolution to be on `Config` kind of discourages people from ignoring that "`resolve()` should be invoked on root config objects, rather than on a subtree", as the Javadoc says.

I ended up exposing a `config()` method returning an unresolved config and left the current behavior of `value()` of forcing resolution. I added appropriate Scaladoc to warn users of the semantics.